# ena-submit
[![Build Status](https://travis-ci.org/rror/ena-submit.svg?branch=master)](https://travis-ci.org/rror/ena-submit)

A Kotlin/Java library to submit [Variant Call Format](https://en.wikipedia.org/wiki/Variant_Call_Format) (VCF) and [CRAM](http://www.ebi.ac.uk/ena/software/cram-toolkit) files to the [European Nucleotide Archive](http://www.ebi.ac.uk/ena) (ENA)

## Submitting a VCF to ENA
[Programmatic submission](http://www.ebi.ac.uk/ena/submit/programmatic-submission) to the European Nucleotide Archive (ENA) requires the creation of 'submission' and 'analysis' XML documents, following the [provided schemas](http://www.ebi.ac.uk/ena/submit/preparing-xmls#submission).
To make this error-prone task easier, the following type-safe builder can be used:

```kotlin
val vcfFile = File("some.vcf")
val (submissionXml: String, analysisXml: String) = analysis {
    alias("Maize HapMap test")
    centerName("CSHL")
    brokerName("ENSEMBL GENOMES")
    holdDate(GregorianCalendar(2020, 11, 24).time) // month is 0 based
    title("Capturing Extant Variation from a Genome in Flux: Maize HapMap II")
    description("A comprehensive characterization of genetic variation across 103 inbred lines ...")
    sampleMapping {
        +"IRGC103469/IRGC103469_aln_sorted.bam".to("SRS302388")
        +"TOG5457/TOG5457_aln_sorted.bam".to("SRS302389")
        +"TOG5467/TOG5467_aln_sorted.bam".to("SRS302390")
    }
    studyReference("SRP011907")
    runReference("SRR447750")
    vcf {
        fileName(vcfFile)  // or fileName("some.vcf")
        md5(vcfFile)  // or md5("10899e2ca49b37c8c37c4763616496ac")
        assemblyAccession("GCA_000005005.2")
        sequenceMapping {
            +"1".to("GK000031.2")
        }
    }
}
```
The order of elements does not matter. Not all elements are mandatory - if any required elements have been omitted an `EnaXmlException` with descriptive message will be thrown.

Now that we have the XMLs, we just have to upload the VCF and submit.
The credentials needed for uploading and submitting to ENA are expected to be stored in the environment variables `ena_user` and `ena_password`.
```kotlin
uploadToEnaFtp(vcfFile)
val result: SubmissionResult = submitToEna(submissionXml, analysisXml, EnaServer.TEST)
if (result.success)
    println("${result.submissionAcc} ${result.analysisAcc}")
else
    println(result.error)
```

## Submitting a VCF to ENA with Java
```java
import kotlin.Pair;
import kotlin.Unit;

File vcf = new File("some.vcf");
Pair<String, String> xmls = GenerateAnalysisKt.analysis(analysis -> {
    analysis.alias("Maize HapMap test");
    analysis.centerName("CSHL");
    analysis.brokerName("ENSEMBL GENOMES");
    analysis.holdDate(new GregorianCalendar(2020, 11, 24).getTime()); // month is 0 based
    analysis.title("Capturing Extant Variation from a Genome in Flux: Maize HapMap II");
    analysis.description("A comprehensive characterization of genetic variation across 103 inbred lines ...");
    analysis.sampleMapping(mapping -> {
        mapping.add("IRGC103469/IRGC103469_aln_sorted.bam", "SRS302388");
        mapping.add("TOG5457/TOG5457_aln_sorted.bam", "SRS302389");
        mapping.add("TOG5467/TOG5467_aln_sorted.bam", "SRS302390");
        return Unit.INSTANCE;
    });
    analysis.studyReference("SRP011907");
    analysis.runReference("SRR447750");
    analysis.vcf(analysisFile -> {
        analysisFile.fileName(vcf);  // or analysisFile.fileName("some.vcf")
        analysisFile.md5(vcf);  // or analysisFile.md5("10899e2ca49b37c8c37c4763616496ac");
        analysisFile.assemblyAccession("GCA_000005005.2");
        analysisFile.sequenceMapping(mapping -> {
            mapping.add("1", "GK000031.2");
            return Unit.INSTANCE;
        });
        return Unit.INSTANCE;
    });
    return Unit.INSTANCE;
});
String submissionXml = xmls.getFirst();
String analysisXml = xmls.getSecond();

SubmitKt.uploadToEnaFtp(vcf);
SubmissionResult result = SubmitKt.submitToEna(submissionXml, analysisXml, EnaServer.TEST);
if (result.getSuccess())
    System.out.println(String.format("%s %s", result.getSubmissionAcc(), result.getAnalysisAcc()));
else
    System.out.println(result.getError());

```

### Generated XML documents
For reference, here are the XML documents generated by the code above:

**submissionXml**
```xml
<SUBMISSION alias="Maize HapMap test" center_name="CSHL" broker_name="ENSEMBL GENOMES">
  <ACTIONS>
    <ACTION>
      <ADD source="analysis.xml" schema="analysis"/>
    </ACTION>
    <ACTION>
      <HOLD HoldUntilDate="2020-12-24Z"/>
    </ACTION>
  </ACTIONS>
</SUBMISSION>
```

**analysisXml**
```xml
<ANALYSIS alias="Maize HapMap test" center_name="CSHL" broker_name="ENSEMBL GENOMES">
  <TITLE>Capturing Extant Variation from a Genome in Flux: Maize HapMap II</TITLE>
  <DESCRIPTION>A comprehensive characterization of genetic variation across 103 inbred lines ...</DESCRIPTION>
  <STUDY_REF accession="SRP011907"/>
  <SAMPLE_REF label="IRGC103469/IRGC103469_aln_sorted.bam" accession="SRS302388"/>
  <SAMPLE_REF label="TOG5457/TOG5457_aln_sorted.bam" accession="SRS302389"/>
  <SAMPLE_REF label="TOG5467/TOG5467_aln_sorted.bam" accession="SRS302390"/>
  <RUN_REF accession="SRR447750"/>
  <ANALYSIS_TYPE>
    <SEQUENCE_VARIATION>
      <ASSEMBLY>
        <STANDARD accession="GCA_000005005.2"/>
      </ASSEMBLY>
      <SEQUENCE accession="GK000031.2" label="1"/>
      <EXPERIMENT_TYPE>Whole genome sequencing</EXPERIMENT_TYPE>
    </SEQUENCE_VARIATION>
  </ANALYSIS_TYPE>
  <FILES>
    <FILE filename="Glab_var_chr1_flt_1k.vcf" filetype="vcf" checksum_method="MD5" checksum="10899e2ca49b37c8c37c4763616496ac"/>
  </FILES>
</ANALYSIS>
```

### Implementation details
XML documents are created with the help of [Apache XMLBeans](https://xmlbeans.apache.org),
which generates an API to match XML schema and compiles it into a *.jar*. The result of this is `libs/ena-schema-1.5.jar`.
Should the [ENA schema](http://www.ebi.ac.uk/ena/submit/read-xml-format-1-5) change, a new version of the API can be generated with:
```bash
xmlbeans-2.6.0/bin/scomp -out ena-schema.jar \
  SRA.analysis.xsd SRA.common.xsd SRA.sample.xsd SRA.study.xsd SRA.submission.xsd
```